src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,376 @@
+    for metadata_file in metadata_files:
+        parent_folder: str = os.path.dirname(metadata_file)
+        package: Optional[str] = get_package_name_from_metadata(metadata_file)
+
+        if package is not None:
+            # Determine which folders or files belong to this package
+            record_file_path = os.path.join(parent_folder, "RECORD")
+            if os.path.exists(record_file_path):
+                # Get unique root folder names
+                with open(record_file_path, encoding="utf-8") as record_file:
+                    record_entries = set()
+
+                    # Read in all record entries
+                    for line in record_file.readlines():
+                        entry = os.path.split(line)[0].split(",")[0]
+                        if entry == "":  # If true, a file present in the root folder
+                            entry = line.split(",")[0]
+                        record_entries.add(entry)
+
+                    # Only select unique base folders or files
+                    included_record_entries = []
+                    for record_entry in record_entries:
+                        record_entry_full_path = os.path.abspath(
+                            os.path.join(directory, record_entry),
+                        )
+                        # RECORD file might contain relative paths to items outside target folder. (ignore these)
+                        if (
+                            os.path.exists(record_entry_full_path)
+                            and directory in record_entry_full_path
+                            and len(record_entry) > 0
+                        ):
+                            included_record_entries.append(record_entry)
+
+                    # Create Requirement objects and store in map
+                    package_name_to_record_entries_map[
+                        Requirement.parse(package)
+                    ] = included_record_entries
+
+    return package_name_to_record_entries_map
+
+
+def identify_supported_packages(
+    packages: List[Requirement],
+    valid_packages: Dict[str, List[str]],
+    native_packages: Set[str],
+) -> Tuple[List[Requirement], List[Requirement], List[Requirement]]:
+    """
+    Detects which `packages` are present in the Snowpark Anaconda channel using the `valid_packages` mapping.
+    If a package is a native dependency (belongs to `native_packages` set) and supported in Anaconda, we switch to
+    the latest available version in Anaconda.
+
+    Note that we also update the `native_packages` set to reflect genuinely problematic native dependencies, i.e.
+    packages that are not present in Anaconda and are likely to cause errors.
+
+    Args:
+        packages (List[Requirement]): List of python packages.
+        valid_packages (Dict[str, List[str]): Mapping from package name to a list of versions available on the Anaconda
+        channel.
+        native_packages (Set[str]): Set of native dependency package names.
+
```

Comment:
```suggestion
        native_packages (Set[str]): Set of package names that have a native dependency.

```
how about this?

Reply:
are they package names requested by the user that we have detected to have a native dependency?

Reply:
Either package names requested by the user (unavailable in anaconda) or dependencies of requested unavailable packages.

Reply:
I see. Could you elaborate the description to cover that?